### PR TITLE
Support multiple periods in line charts and hover tooltip

### PR DESCRIPTION
**Pull request by olslash**
Definitely going to need code review on this-- it allows adding periods onto the individual metrics that are passed into the line chart, and in that case, overlays the lines on top of each other. I also updated the hover mixin to support using these metric periods instead of the single chart-wide period.

Still needs a little cleanup work which I'm going to get at tomorrow.

### CODE REVIEW

File: docs/examples/widgets/line.html [hunk 1]
```diff
@@ -105,6 +105,79 @@
               organizationId: "54c22ac91342df12d0f2a288"
             });
```
Comment (Juice10) [inline docs/examples/widgets/line.html]: No need to keep the `Openhose.configure` above, as you are providing the stream id/org ids per metric.

File: src/processor/analytics-mixin.js [hunk 2]
```diff
@@ -90,111 +36,116 @@ define(['../core', '../lib/utils'], function(Openhose, Utils) {
         });
       }.bind(this));
 
-      return this._alignSegments(dimensions[0].metrics, segments);
+     return this._alignSegments(dimensions[0].metrics, segments);
```
Comment (Juice10) [inline src/processor/analytics-mixin.js]: Probably just a mistake but...



:wink:

File: src/processor/analytics-mixin.js [hunk 3]
```diff
@@ -90,111 +36,116 @@ define(['../core', '../lib/utils'], function(Openhose, Utils) {
...
     },
 
+    hasDifferentPeriods: function(metrics) {
+
+      var periodAttrs = metrics.map(function(metric) {
+        return metric.period.attributes;
```
Comment (Juice10) [inline src/processor/analytics-mixin.js]: You probably want to use `.toJSON()` instead of `.attributes`

File: src/processor/analytics-mixin.js [hunk 4]
```diff
@@ -90,111 +36,116 @@ define(['../core', '../lib/utils'], function(Openhose, Utils) {
...
+      });
+
+      var allSamePeriod = !!periodAttrs.reduce(function(prev, curr) {
+        if(prev === false) return false;
+
+        if(curr.start === prev.start && curr.end === prev.end) {
```
Comment (Juice10) [inline src/processor/analytics-mixin.js]: You should also check the bucket size as that can influence the data returned for a period heavily

File: src/processor/analytics-mixin.js [hunk 5]
```diff
@@ -90,111 +36,116 @@ define(['../core', '../lib/utils'], function(Openhose, Utils) {
...
+
+      var anyPreviousPeriodFlags = metrics.some(function(metric) {
+        if (metric.get('previousPeriod')) {
+          return true;
+        }
+      });
+
+
+      if (anyPreviousPeriodFlags) {
+        return false;
```
Comment (Juice10) [inline src/processor/analytics-mixin.js]: Shouldn't this be `return true`?

File: src/processor/analytics-mixin.js [hunk 6]
```diff
@@ -90,111 +36,116 @@ define(['../core', '../lib/utils'], function(Openhose, Utils) {
...
-          var currentTs = Utils.fromISO(metrics[i - 1].collection.at(0).get('timestamp'));
-          segment.timeDelta = currentTs - prevTs;
+          values = doOffsetOneBucket ? this._offsetByOneBucket(values, metric.period) : values;
+          var segment = {
+            label: metric.getLabel({context: options.context}),
+            period: metric.period, // TODO: do this for the other segments (above) too
+            values: values,
+            relativeValues: metric.relativeValues
+          };
+          segment.timeDelta = prevTs - startsAt;
```
Comment (Juice10) [inline src/processor/analytics-mixin.js]: You definitely need to test what happens when you mix daylight savings time, non daylight savings and daylight saving transition time buckets

File: src/processor/analytics-mixin.js [hunk 7]
```diff
@@ -90,111 +36,116 @@ define(['../core', '../lib/utils'], function(Openhose, Utils) {
...
+        metrics.forEach(function (metric, i) {
 
-          var moments = metric.period.toPreviousMoments();
-          segment.label = segment.label + ' ' + Utils.getShortTimeRangeLabel(moments.start, moments.end);
-        }
+          var tl = metric.timeline();
 
-        segments.push(segment);
-      }.bind(this));
+          var values = this._fillMissingGaps(metric.period, tl);
```
Comment (Juice10) [inline src/processor/analytics-mixin.js]: You can leave `_fillMissingGaps` out, its deprecated,
they just return the second value you provided them.

File: src/processor/line.js [hunk 8]
```diff
@@ -51,16 +56,19 @@ define([
...
       var timeDelta = 0;
       var period = graph.period;
+
       if (graph.isHistory) {
         timeDelta = graph.timeDelta;
         // and set prev x value to be overlapping with current
-        graphValues.map(function(x) {
-          x.x = graph.period.incrementMomentByPeriod(x.x * 1000).valueOf() / 1000;
+        graphValues.forEach(function(x, i) {
+          x.x = graphValuesFirst[i].x;
```
Comment (Juice10) [inline src/processor/line.js]: any reason why you removed `graph.period.incrementMomentByPeriod`? it seemed to fix a lot of issues.
Would also be good to write a unit test for this method to list what kind of data you expect when processing multiple periods in one graph.

File: src/processor/line.js [hunk 9]
```diff
@@ -121,13 +130,12 @@ define([
     }.bind(this));
 
     if (!this.options.noZeroFill) {
-      AnalyticsMixin.zeroFill(lines);
+      //AnalyticsMixin.zeroFill(lines);
```
Comment (Juice10) [inline src/processor/line.js]: What was your reason for removing the zero filling?
And if you really want to remove it you should also remove the line above it and everywhere else where `noZeroFill` is used

File: src/visualization/hover-mixin.js [hunk 10]
```diff
@@ -95,22 +95,35 @@ define(['../core', './lib/custom-hover-details', '../lib/utils'], function(Openh
...
+                }
 
-              // revert the offset.
-              if (this.doOffsetOneBucket) {
-                d = moment(dTs - this.mainPeriod.getBucketMs()).tz(period.timezone());
-              } else {
-                d = moment(dTs).tz(period.timezone());
+                var displayTime;
+
+//              var d;
```
Comment (Juice10) [inline src/visualization/hover-mixin.js]: you can get rid of this line

File: src/visualization/hover-mixin.js [hunk 11]
```diff
@@ -95,22 +95,35 @@ define(['../core', './lib/custom-hover-details', '../lib/utils'], function(Openh
...
+                var d2;
+
+//              // revert the offset.
```
Comment (Juice10) [inline src/visualization/hover-mixin.js]: you can get rid of the doubled comments

**Comment by Juice10** [comment]
Please don't commit the bundled files. They make merging hell.
All in all not bad for a first crack at the Vis SDK

File: src/processor/analytics-mixin.js (hunk 7 above)
Comment (drakonen) [inline src/processor/analytics-mixin.js]: clean up opportunity

File: src/processor/line.js (hunk 9 above)
Comment (drakonen) [inline src/processor/line.js]: I tihnk zerofill is an empty function. if that is the case, remove it.

**Comment by drakonen** [comment]
Looks good, Justin pointed out a few issues, get those fixed and I think we're go.

Add some tests to see how it works cross daylight savings.

File: docs/examples/widgets/line.html (hunk 1 above)
Comment (olslash) [inline docs/examples/widgets/line.html]: :heavy_check_mark:

File: src/processor/analytics-mixin.js (hunk 2 above)
Comment (olslash) [inline src/processor/analytics-mixin.js]: whoops :heavy_check_mark:

File: src/processor/analytics-mixin.js (hunk 3 above)
Comment (olslash) [inline src/processor/analytics-mixin.js]: not sure there's a compelling reason to call toJSON unless you need the data to be copied-- in this case attributes is equivalent

File: src/processor/analytics-mixin.js (hunk 5 above)
Comment (olslash) [inline src/processor/analytics-mixin.js]: The intent is to find lines that have attached periods, but aren't flagged as previous period, as that would be handled by the existing code. Maybe hasDifferentPeriods isn't a descriptive enough name, i'll change it

File: src/processor/analytics-mixin.js (hunk 7 above)
Comment (olslash) [inline src/processor/analytics-mixin.js]: :heavy_check_mark:

File: src/processor/line.js (hunk 8 above)
Comment (olslash) [inline src/processor/line.js]: part of kaare's changes-- trying to figure it out :) What issues does it fix, so I know if I need it?

File: src/processor/line.js (hunk 9 above)
Comment (olslash) [inline src/processor/line.js]: The function isn't empty-- but I think it was only commented out for testing, just something I missed when cleaning up.

File: src/visualization/hover-mixin.js (hunk 11 above)
Comment (olslash) [inline src/visualization/hover-mixin.js]: :heavy_check_mark:

**Comment by olslash** [comment]
Regarding the bundles, I can go back when this is solid and rebase the commits to remove them. Sorry, didn't realize.

File: src/processor/analytics-mixin.js (hunk 6 above)
Comment (olslash) [inline src/processor/analytics-mixin.js]: What would a failing case for DST be? The behavior right now is that graphs render correctly, and the hover tooltip simply skips the non-existant hour that's skipped by DST-- the only possible issue is that the hover tooltip period duration can be 1 hour different before and after the inflection point.

For example: DST shift happens March 8 at 2am this year-- the hover goes from 
<March 8, 12am - 10am>
<March 8, 1am - 11am>
----shift---
<March 8, 3am - 12pm>  (we skipped 2am, and lost one hour on the duration-- from 10 hrs to 9)


If two lines in the graph have DST shifts at different points, we get the same behavior with nothing unexpected.

**Comment by olslash** [comment]
@kaareal @kevinvandijk  this is ready to go with your :+1:

**Comment by olslash** [comment]
Made some additonal commits to this that should bring it really close to ready. Staging streams have been down all day so I can't verify that everything's working in NC, but my example code's working well. Would appreciate another go-over

**Comment by drakonen** [comment]
New changes look fine. Be sure to address the other things pointed out before.

**Comment by drakonen** [comment]
Also, isTemporal and isHistory seems double. Maybe we can fix that?

File: src/processor/analytics-mixin.js (hunk 6 above)
Comment (olslash) [inline src/processor/analytics-mixin.js]: @drakonen @Juice10 any idea about this one?

File: src/processor/line.js (hunk 8 above)
Comment (olslash) [inline src/processor/line.js]: @drakonen @Juice10 any idea about this one?

**Comment by olslash** [comment]
The problem is that isHistory has other implications-- ones that temporal takes advantage of, but also for lots of other unrelated stuff. I tried a bunch of approaches, but ultimately needed another flag to keep from breaking existing functionality. If you can think of a way around it, i'd love to hear.

File: src/processor/analytics-mixin.js (hunk 6 above)
Comment (drakonen) [inline src/processor/analytics-mixin.js]: Do it as it is now, we can fix this later. Lets get this out of the way.

File: src/processor/line.js (hunk 8 above)
Comment (drakonen) [inline src/processor/line.js]: This is good.
You're setting the x of the other segments to be like the first segment.

**Comment by drakonen** [comment]
Re isHistory/isTemporal
I think we should remove the isHistory in favor of the new way (but still provide the higher level shortcut in the widget).
But not now, we should do it in a future refactoring.

**Pull request merged**